README.md
# AT24CXX 软件包

## 1 介绍

AT24CXX 软件包提供了at24cxx 系列 EEPROM 基本功能。本文介绍该软件包的基本读写功能，以及 `Finsh/MSH` 测试命令等。
目前已在 stm32l475-atk-pandora 上验证 at24c02 通过。

### 1.1 目录结构

| 名称 | 说明 |
| ---- | ---- |
| at24cxx.h | EEPROM 使用头文件 |
| at24cxx.c | EEPROM 使用源代码 |
| SConscript | RT-Thread 默认的构建脚本 |
| README.md | 软件包使用说明 |
| at24cxx_datasheet.pdf | 官方数据手册 |

### 1.2 许可证

AT24CXX 软件包遵循  Apache-2.0 许可，详见 LICENSE 文件。

### 1.3 依赖

依赖 `RT-Thread I2C` 设备驱动框架。

## 2 获取软件包

使用 `at24cxx` 软件包需要在 RT-Thread 的包管理器中选择它，具体路径如下：

```
RT-Thread online packages
    peripheral libraries and drivers  --->
        [*] at24cxx: eeprom at24cxx driver library  --->
            Version (latest)  --->
```


每个功能的配置说明如下：

- `at24cxx: ee2prom at24cxx driver library`：选择使用 `at24cxx` 软件包；
- `Version`：配置软件包版本，默认最新版本。

然后让 RT-Thread 的包管理器自动更新，或者使用 `pkgs --update` 命令更新包到 BSP 中。

## 3 使用 at24cxx 软件包

按照前文介绍，获取 `at24cxx` 软件包后，就可以按照 下文提供的 API 使用 ee2prom `at24cxx 与 `Finsh/MSH` 命令进行测试，详细内容如下。

### 3.1 API

#### 3.1.1  初始化 

`at24cxx_device_t at24cxx_init(const char *i2c_bus_name)`

根据总线名称，自动初始化对应的 AT24CXX 设备，具体参数与返回说明如下表

| 参数    | 描述                      |
| :----- | :----------------------- |
| name   | i2c 设备名称 |
| **返回** | **描述** |
| != NULL | 将返回 at24cxx 设备对象 |
| = NULL | 查找失败 |

#### 3.1.2  反初始化

void at24cxx_deinit(aht10_device_t dev)

如果设备不再使用，反初始化将回收 at24cxx 设备的相关资源，具体参数说明如下表

| 参数 | 描述           |
| :--- | :------------- |
| dev  | at24cxx 设备对象 |

#### 3.1.3 读取 

rt_err_t at24cxx_read(struct rt_i2c_bus_device *bus,uint8_t ReadAddr,uint8_t *pBuffer,uint16_t NumToRead)

通过 `at24cxx` 读取 eeprom ，在AT24CXX里面的指定地址开始读出指定个数的数据，具体参数与返回说明如下表

| 参数     | 描述           |
| :------- | :------------- |
| bus      | at24cxx 设备对象 |
| **返回** | **描述**       |
| != RT_EOK | 读取成功     |
| = RT_EOK| 读取失败       |

#### 3.1.4 写入

rt_err_t at24cxx_write(struct rt_i2c_bus_device *bus,uint8_t WriteAddr,uint8_t *pBuffer,uint16_t NumToWrite)


通过 `at24cxx` 写入，在AT24CXX里面的指定地址开始写入指定个数的数据，具体参数与返回说明如下表

| 参数     | 描述          |
| :------- | :------------ |
| bus      | at24cxx 设备对象 |
| **返回** | **描述**      |
| != RT_EOK  | 写入失败    |
| =RT_EOK    | 写入成功      |

#### 3.1.5 查看设备

rt_err_t at24cxx_check(struct rt_i2c_bus_device *bus)

查看是否存在eeprom 设备，通过在最后一个字节写入标志位，进行判断，具体参数与返回说明如下表

| 参数     | 描述          |
| :------- | :------------ |
| bus      | at24cxx 设备对象 |
| **返回** | **描述**      |
| != RT_EOK  | 设备不存在    |
| =RT_EOK    | 设备存在      |


### 3.2 Finsh/MSH 测试命令

at24cxx 软件包提供了丰富的测试命令，项目只要在 RT-Thread 上开启 Finsh/MSH 功能即可。在做一些基于 `at24cxx` 的应用开发、调试时，这些命令会非常实用，它可以准确的读取指传感器测量的温度与湿度。具体功能可以输入 `at24cxx` ，可以查看完整的命令列表

```
msh />at24cxx
Usage:
at24cxx probe <dev_name>   - probe sensor by given name
at24cxx check              - check device
at24cxx write              - write  data
at24cxx read               - read   data
msh />
```

#### 3.2.1 在指定的 i2c 总线上探测传感器 

当第一次使用 `at24cxx` 命令时，直接输入 `at24cxx probe <dev_name>` ，其中 `<dev_name>` 为指定的 i2c 总线，例如：i2c0。如果有这个设备，就不会提示错误；如果总线上没有这个设备，将会显示提示找不到相关设备，日志如下：

```
msh />at24cxx probe i2c1      #探测成功，没有错误日志
msh />
msh />at24cxx probe i2c88     #探测失败，提示对应的 I2C 设备找不到
[E/aht10] can't find at24cxx device on 'i2c88'
msh />
```

## 4 注意事项

- 目前软件包在连续写入多个字节时候，中间需要适当的延时，才能保证写入成功，程序目前是延迟 1 ms 时间 。
- 从设备地址为7位地址 0x50, 而不是 0xA0 。

## 5 联系方式

* 维护：[FanXiaojie](https://github.com/XiaojieFan)
* 主页：https://github.com/XiaojieFan/at24cxx



ChangeLog.md
1.fix bug:if write addr is not 0,write number is wrong

2.fix bug:if write number is over 255,write number is wrong

3.add 16bit addr support


at24cxx.c
/*
 * Copyright (c) 2006-2018, RT-Thread Development Team
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Change Logs:
 * Date           Author       Notes
 * 2019-04-13     XiaojieFan   the first version
 */
#include <rtthread.h>
#include <rtdevice.h>
#include <board.h>


#include <string.h>

#define DBG_ENABLE
#define DBG_SECTION_NAME "AHT10"
#define DBG_LEVEL DBG_LOG
#define DBG_COLOR
#include <rtdbg.h>

#include "at24cxx.h"

#ifdef PKG_USING_AT24CXX
#define AT24CXX_ADDR 0x50                      //A0 A1 A2 connect GND

static rt_err_t read_regs(struct rt_i2c_bus_device *bus, rt_uint8_t len, rt_uint8_t *buf)
{
    struct rt_i2c_msg msgs;

    msgs.addr = AT24CXX_ADDR;
    msgs.flags = RT_I2C_RD;
    msgs.buf = buf;
    msgs.len = len;

    if (rt_i2c_transfer(bus, &msgs, 1) == 1)
    {
        return RT_EOK;
    }
    else
    {
        return -RT_ERROR;
    }
}
uint8_t at24cxx_read_one_byte(struct rt_i2c_bus_device *bus, uint16_t readAddr)
{
    rt_uint8_t buf[2];
    rt_uint8_t temp;
#if	(EE_TYPE > AT24C16)  
    buf[0] = (uint8_t)(readAddr>>8);	
    buf[1] = (uint8_t)readAddr;
    if (rt_i2c_master_send(bus, AT24CXX_ADDR, 0, buf, 2) == 0) 
#else
    buf[0] = readAddr;
    if (rt_i2c_master_send(bus, AT24CXX_ADDR, 0, buf, 1) == 0)
#endif        
    {
        return -RT_ERROR;
    }
    read_regs(bus, 1, &temp);
    return temp;
}

rt_err_t at24cxx_write_one_byte(struct rt_i2c_bus_device *bus, uint16_t writeAddr, uint8_t dataToWrite)
{
    rt_uint8_t buf[3];
#if	(EE_TYPE > AT24C16)      
    buf[0] = (uint8_t)(writeAddr>>8);	
    buf[1] = (uint8_t)writeAddr;
    buf[2] = dataToWrite;
    if (rt_i2c_master_send(bus, AT24CXX_ADDR, 0, buf, 3) == 3)    
#else    
    buf[0] = writeAddr; //cmd
    buf[1] = dataToWrite;
    //buf[2] = data[1];
    

    if (rt_i2c_master_send(bus, AT24CXX_ADDR, 0, buf, 2) == 2)
#endif        
        return RT_EOK;
    else
        return -RT_ERROR;

}

rt_err_t at24cxx_check(at24cxx_device_t dev)
{
    uint8_t temp;
    rt_err_t result;
    RT_ASSERT(dev);

    temp = at24cxx_read_one_byte(dev->i2c, 255);
    if (temp == 0x55) return RT_EOK;
    else
    {
        at24cxx_write_one_byte(dev->i2c, 255, 0x55);
        temp = at24cxx_read_one_byte(dev->i2c, 255);
        if (temp == 0x55) return RT_EOK;
    }
    return RT_ERROR;
}

/**
 * This function read the specific numbers of data to the specific position
 *
 * @param bus the name of at24cxx device
 * @param ReadAddr the start position to read
 * @param pBuffer  the read data store position
 * @param NumToRead
 * @return RT_EOK  write ok.
 */
rt_err_t at24cxx_read(at24cxx_device_t dev, uint16_t ReadAddr, uint8_t *pBuffer, uint16_t NumToRead)
{
    rt_err_t result;
    RT_ASSERT(dev);
    result = rt_mutex_take(dev->lock, RT_WAITING_FOREVER);
    if (result == RT_EOK)
    {
        while (NumToRead)
        {
            *pBuffer++ = at24cxx_read_one_byte(dev->i2c, ReadAddr++);
            NumToRead--;
        }
    }
    else
    {
        LOG_E("The at24cxx could not respond  at this time. Please try again");
    }
    rt_mutex_release(dev->lock);

    return RT_EOK;
}

/**
 * This function write the specific numbers of data to the specific position
 *
 * @param bus the name of at24cxx device
 * @param WriteAddr the start position to write
 * @param pBuffer  the data need to write
 * @param NumToWrite
 * @return RT_EOK  write ok.at24cxx_device_t dev
 */
rt_err_t at24cxx_write(at24cxx_device_t dev, uint16_t WriteAddr, uint8_t *pBuffer, uint16_t NumToWrite)
{
    uint16_t i = 0;
    rt_err_t result;
    RT_ASSERT(dev);
    result = rt_mutex_take(dev->lock, RT_WAITING_FOREVER);
    if (result == RT_EOK)
    {
        while (1) //NumToWrite--
        {
            if (at24cxx_write_one_byte(dev->i2c, WriteAddr, pBuffer[i]) != RT_EOK)
            {
                rt_thread_mdelay(EE_TWR);
            }
            else
            {
                WriteAddr++;
                i++;
            }
            if (i == NumToWrite)
            {
                break;
            }

        }
    }
    else
    {
        LOG_E("The at24cxx could not respond  at this time. Please try again");
    }
    rt_mutex_release(dev->lock);

    return RT_EOK;
}
/**
 * This function initializes at24cxx registered device driver
 *
 * @param dev the name of at24cxx device
 *
 * @return the at24cxx device.
 */
at24cxx_device_t at24cxx_init(const char *i2c_bus_name)
{
    at24cxx_device_t dev;

    RT_ASSERT(i2c_bus_name);

    dev = rt_calloc(1, sizeof(struct at24cxx_device));
    if (dev == RT_NULL)
    {
        LOG_E("Can't allocate memory for at24cxx device on '%s' ", i2c_bus_name);
        return RT_NULL;
    }

    dev->i2c = rt_i2c_bus_device_find(i2c_bus_name);
    if (dev->i2c == RT_NULL)
    {
        LOG_E("Can't find at24cxx device on '%s' ", i2c_bus_name);
        rt_free(dev);
        return RT_NULL;
    }

    dev->lock = rt_mutex_create("mutex_at24cxx", RT_IPC_FLAG_FIFO);
    if (dev->lock == RT_NULL)
    {
        LOG_E("Can't create mutex for at24cxx device on '%s' ", i2c_bus_name);
        rt_free(dev);
        return RT_NULL;
    }

    return dev;
}

/**
 * This function releases memory and deletes mutex lock
 *
 * @param dev the pointer of device driver structure
 */
void at24cxx_deinit(at24cxx_device_t dev)
{
    RT_ASSERT(dev);

    rt_mutex_delete(dev->lock);

    rt_free(dev);
}

uint8_t TEST_BUFFER[] = "WELCOM TO RTT";
#define SIZE sizeof(TEST_BUFFER)

void at24cxx(int argc, char *argv[])
{
    static at24cxx_device_t dev = RT_NULL;

    if (argc > 1)
    {
        if (!strcmp(argv[1], "probe"))
        {
            if (argc > 2)
            {
                /* initialize the sensor when first probe */
                if (!dev || strcmp(dev->i2c->parent.parent.name, argv[2]))
                {
                    /* deinit the old device */
                    if (dev)
                    {
                        at24cxx_deinit(dev);
                    }
                    dev = at24cxx_init(argv[2]);
                }
            }
            else
            {
                rt_kprintf("at24cxx probe <dev_name>   - probe sensor by given name\n");
            }
        }
        else if (!strcmp(argv[1], "read"))
        {
            if (dev)
            {
                uint8_t testbuffer[50];

                /* read the eeprom data */
                at24cxx_read(dev, 0, testbuffer, SIZE);

                rt_kprintf("read at24cxx : %s\n", testbuffer);

            }
            else
            {
                rt_kprintf("Please using 'at24cxx probe <dev_name>' first\n");
            }
        }
        else if (!strcmp(argv[1], "write"))
        {
            at24cxx_write(dev, 0, TEST_BUFFER, SIZE);
            rt_kprintf("write ok\n");
        }
        else if (!strcmp(argv[1], "check"))
        {
            if (at24cxx_check(dev) == 1)
            {
                rt_kprintf("check faild \n");
            }
        }
        else
        {
            rt_kprintf("Unknown command. Please enter 'at24cxx0' for help\n");
        }
    }
    else
    {
        rt_kprintf("Usage:\n");
        rt_kprintf("at24cxx probe <dev_name>   - probe eeprom by given name\n");
        rt_kprintf("at24cxx check              - check eeprom at24cxx \n");
        rt_kprintf("at24cxx read               - read eeprom at24cxx data\n");
        rt_kprintf("at24cxx write              - write eeprom at24cxx data\n");

    }
}
MSH_CMD_EXPORT(at24cxx, at24cxx eeprom function);
#endif
